### DISCUSSION TITLE
Extending Types BaseClass => ChildClass

### DISCUSSION THREAD

**Discussion by zeroedin** [RedHat-UX/red-hat-design-system#658] [General]
What is the best solution for extending types when different options are needed in separate child classes?

My original attempt was thinking I could just override the `InsetVariant` type in `rh-tabs` so I would not have to reinclude `@property` in either `pfe-tabs` or `rh-tabs` that however did not work.
 
Here is a proposal @heyMP and I worked out. 

If anyone else has any other thoughts please let me know.

**BaseTabs.ts**
```ts
export type InsetVariant = (
  | 'sm'
  | 'md'
  | 'lg'
  | 'xl'
  | '2xl'
  | 'page'
);

...

@property({ reflect: false }) inset?: InsetVariant;
```

**pfe-tabs.ts**
```ts
import type { InsetVariant } from './BaseTabs.js'
export type ExtendedInsetVariant = InsetVariant | 'page'

...

@property({ reflect: false }) inset?: ExtendedInsetVariant;
```

**rh-tabs.ts**
```ts
import type { InsetVariant } from '@patternfly/pfe-tabs/BaseTab.js'
export type ExtendedInsetVariant = InsetVariant | 'none'

...

@property({ reflect: false }) inset?: ExtendedInsetVariant;
```

**Comment by bennypowers** [comment]
Today at Office Hours it was resolved that
1. Base classes should be as abstract as possible. (I will elaborate on this in the wiki)
2. Class Inheritance should be a feature of design discussions, particularly focusing on templates and/or template hooks in base classes. 

Thank you @zeroedin  for raising the discussion, @marionnegp  @heyMP , @nikkimk  @brianferry  for your input.

**Reply by bennypowers** [reply]
Resulting Wiki entry: https://github.com/RedHat-UX/red-hat-design-system/wiki/Base-Classes

**Comment by zeroedin** [comment]
Initial work on BaseClasses for tabs in PFE:  https://github.com/patternfly/patternfly-elements/pull/2260
